rust_src/src/dispnew.rs

```diff
@@ -136,4 +143,57 @@ pub fn internal_show_cursor_p(window: LispWindowOrSelected) -> bool {
     !win.cursor_off_p()
 }
 
+/// Beep, or flash the screen.
+/// Also, unless an argument is given,
+/// terminate any keyboard macro currently executing.
+#[lisp_fn(min = "0")]
+pub fn ding(arg: LispObject) {
+    if arg.is_nil() {
+        unsafe { bitch_at_user() }
+    } else {
+        unsafe {
+            if noninteractive {
+                putchar_unlocked(7);
+            } else {
+                ring_bell(selected_frame().as_mut())
+            }
+        }
+    }
+}
+
```

Comment:
bitch_at_user is a version of `ding` that sometimes signals.....

```
void
bitch_at_user (void)
{
  if (noninteractive)
    putchar_unlocked (07);
  else if (!INTERACTIVE)  /* Stop executing a keyboard macro.  */
    {
      const char *msg
	= "Keyboard macro terminated by a command ringing the bell";
      Fsignal (Quser_error, list1 (build_string (msg)));
    }
  else
    ring_bell (XFRAME (selected_frame));
}
```

Reply:
It seems to me that `ding` would be written:

```
  if (noninteractive)
    putchar_unlocked (07);
  else if (NILP(arg) && !INTERACTIVE)  /* Stop executing a keyboard macro.  */
    {
      const char *msg
	= "Keyboard macro terminated by a command ringing the bell";
      Fsignal (Quser_error, list1 (build_string (msg)));
    }
  else
    ring_bell (XFRAME (selected_frame));
}
```

Reply:
I'd port `bitch_at_user` as above and call it `ding_internal`. Have `ding` and `bitch_at_user` call it and `bitch_at_user` always passes in NIL.

Reply:
As an aside, that is an archaic name for a function that we should consider retiring.